### PR TITLE
(BREAKING) O3-2748 service queues - add ability to undo transitions and void queue entries

### PR DESCRIPTION
**Pull request by chibongho**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing.en-US#contributing-guidelines) label. See existing PR titles for inspiration.
- [x] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://om.rs/o3ui).
- [x] My work includes tests or is validated by existing tests.

## Summary

This PR adds additions actions on queue entries to be able to void them or undo transitions in case the entry is created / transitioned by mistake.

This PR is dependent on cooresponding change to the backend. [PR](https://github.com/openmrs/openmrs-module-queue/pull/64) here.

## Screenshots

image
image
image

### CODE REVIEW

**Comment by ojwanganto** [comment]
This is a good addition @chibongho. I am wondering if this can be extended to take care of queuing back to the previous location i.e. from lab to clinical room. Do you have such use-cases in your workflows?

**Review by ojwanganto — APPROVED** [review]
LGTM

**Comment by chibongho** [comment]
> This is a good addition @chibongho. I am wondering if this can be extended to take care of queuing back to the previous location i.e. from lab to clinical room. Do you have such use-cases in your workflows?

@ojwanganto with the current implementation, undoing a transition will only queue the patient back to wherever their previous queue entry was. If their most recent transition was from a different location, it would do what you described. Otherwise, the user will have to do "undo transition" multiple times until the system re-activates a queue entry from a previous location. 

In our use case, we only have one queue per location, so the distinction between queue and location is not important to us. Do you have a use case for undo multiple transitions at once?

File: packages/esm-service-queues-app/src/queue-table/cells/queue-table-action-cell.component.tsx [hunk 1]
```diff
@@ -9,6 +9,31 @@ export function QueueTableActionCell({ queueEntry }: QueueTableCellComponentProp
 
   return (
     <div>
+      {queueEntry.previousQueueEntry == null ? (
+        <Button
+          kind="ghost"
+          aria-label="Actions"
```
Comment (denniskigen) [inline packages/esm-service-queues-app/src/queue-table/cells/queue-table-action-cell.component.tsx:15]: Nit: this should be translated

File: packages/esm-service-queues-app/src/queue-table/cells/queue-table-action-cell.component.tsx [hunk 2]
```diff
@@ -9,6 +9,31 @@ export function QueueTableActionCell({ queueEntry }: QueueTableCellComponentProp
...
+              closeModal: () => dispose(),
+              queueEntry,
+            });
+          }}>
+          {t('void', 'Void')}
+        </Button>
+      ) : (
+        <Button
+          kind="ghost"
+          aria-label="Actions"
```
Comment (denniskigen) [inline packages/esm-service-queues-app/src/queue-table/cells/queue-table-action-cell.component.tsx:27]: Same here

File: packages/esm-service-queues-app/src/queue-table/queue-entry-actions/queue-entry-actions-modal.component.tsx [hunk 3]
```diff
@@ -47,7 +47,11 @@ interface FormParams {
 }
 
 // Modal with form to provide the same UI for editting or transitioning a queue entry
```
Comment (denniskigen) [inline packages/esm-service-queues-app/src/queue-table/queue-entry-actions/queue-entry-actions-modal.component.tsx:49]:
```suggestion
// Modal with a form to provide the same UI for editing or transitioning a queue entry
```

**Review by denniskigen — APPROVED** [review]
Nice work, @chibongho! I've left a couple of nitpicky suggestions.

**Review by CynthiaKamau — APPROVED** [review]

File: packages/esm-service-queues-app/src/queue-table/cells/queue-table-action-cell.component.tsx [hunk 4]
```diff
@@ -9,6 +9,31 @@ export function QueueTableActionCell({ queueEntry }: QueueTableCellComponentProp
...
+          onClick={() => {
+            const dispose = showModal('void-queue-entry-modal', {
+              closeModal: () => dispose(),
+              queueEntry,
+            });
+          }}>
+          {t('void', 'Void')}
```
Comment (mseaton) [inline packages/esm-service-queues-app/src/queue-table/cells/queue-table-action-cell.component.tsx:22]: "Void" isn't really an end-user facing term that we tend to use.  Typically, we'd present this as this "Delete".  Voiding is a technical detail as to how we perform deletes on the back end.

File: packages/esm-service-queues-app/src/queue-table/cells/queue-table-action-cell.component.tsx [hunk 5]
```diff
@@ -9,6 +9,31 @@ export function QueueTableActionCell({ queueEntry }: QueueTableCellComponentProp
...
+        <Button
+          kind="ghost"
+          aria-label="Actions"
+          onClick={() => {
+            const dispose = showModal('undo-transition-queue-entry-modal', {
+              closeModal: () => dispose(),
+              queueEntry,
+            });
+          }}>
+          {t('moveBack', 'Move back')}
```
Comment (mseaton) [inline packages/esm-service-queues-app/src/queue-table/cells/queue-table-action-cell.component.tsx:34]: This isn't really correct.  "Move back" to me, implies that the original transition was correct, and now they are done there and need to go back in the other queue again.  This is really here as an _alternative_ to "Delete", which is essentially "Delete and reactivate previous queue entry" or "Undo this transition".  We need to make it explicit that this is here to _revert_ the previous action, as if it had never happened.

File: packages/esm-service-queues-app/src/queue-table/queue-entry-actions/queue-entry-undo-actions-modal.component.tsx [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+          closeModal();
+        } else {
+          throw { message: t('unexpectedServerResponse', 'Unexpected Server Response') };
+        }
+      })
+      .catch((error) => {
+        showSnackbar({
+          title: submitFailureTitle,
+          kind: 'error',
+          subtitle: error?.message,
```
Comment (mseaton) [inline packages/esm-service-queues-app/src/queue-table/queue-entry-actions/queue-entry-undo-actions-modal.component.tsx:67]: Ideally you'd work with your back end changes to ensure whatever message is coming from the back end that indicates a failure state (eg. no previous queue entry, multiple matching previous queue entries, etc) could be passed here via a localized message.  i.e. can we create an appropriate Exception class that, when it propogates through to here, we have access to both a message code and a default message that we can use to present the error.

File: packages/esm-service-queues-app/src/queue-table/queue-entry-actions/void-queue-entry-modal.component.tsx [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+interface VoidQueueEntryModalProps {
+  queueEntry: QueueEntry;
+  closeModal: () => void;
+}
+
+const VoidQueueEntryModal: React.FC<VoidQueueEntryModalProps> = ({ queueEntry, closeModal }) => {
+  const { t } = useTranslation();
+
+  const modalInstruction = <p>{t('confirmVoidQueueEntry', 'Are you sure you want to void this queue entry?')}</p>;
```
Comment (mseaton) [inline packages/esm-service-queues-app/src/queue-table/queue-entry-actions/void-queue-entry-modal.component.tsx:15]: void -> delete

File: packages/esm-service-queues-app/src/queue-table/queue-entry-actions/void-queue-entry-modal.component.tsx [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+      queueEntry={queueEntry}
+      closeModal={closeModal}
+      modalParams={{
+        modalTitle: t('voidQueueEntry', 'Void queue entry'),
+        modalInstruction,
+        submitButtonText: t('voidQueueEntry', 'Void queue entry'),
+        submitSuccessTitle: t('queueEntryVoidSuccessful', 'Queue entry voided successfully'),
+        submitSuccessText: t('queueEntryVoidSuccessful', 'Queue entry voided successfully'),
+        submitFailureTitle: t('queueEntryVoidFailed', 'Error voiding queue entry'),
+        submitAction: (queueEntry) => voidQueueEntry(queueEntry.uuid),
```
Comment (mseaton) [inline packages/esm-service-queues-app/src/queue-table/queue-entry-actions/void-queue-entry-modal.component.tsx:28]: Void -> Delete above where appropriate.

**Review by mseaton — CHANGES_REQUESTED** [review]
Overall looks great @chibongho , nice work!  See my comments for some changes requested, nothing major.

**Review by mseaton — APPROVED** [review]
LGTM - thanks!

**Comment by sam-sny** [comment]
Good One

**Pull request merged**